core/src/impl/Kokkos_ViewHooks.hpp

```diff
@@ -0,0 +1,466 @@
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_VIEWHOOKS_HPP
+#define KOKKOS_VIEWHOOKS_HPP
+
+#include <Kokkos_Core_fwd.hpp>
+#include <functional>
+#include <memory>
+#include <type_traits>
+#include <map>
+#include <string>
+
+// default implementation for view holder specialization
+namespace Kokkos {
+namespace Experimental {
+namespace Impl {
+
+template <class ViewType, class Enable = void>
+class ViewHookDeepCopy;
+
+template <class ViewType, class Enable = void>
+class ViewHookUpdate {
+ public:
+  static inline void update_view(ViewType &, const void *) {}
+  static constexpr const char *m_name = "ConstDefault";
+};
+
+template <class ViewType, class Enable = void>
+class ViewHookCopyView {
+ public:
+  static inline void copy_view(ViewType &, const void *) {}
+  static inline void copy_view(const void *, ViewType &) {}
+  static constexpr const char *m_name = "ConstDefault";
+};
+
+}  // namespace Impl
+}  // namespace Experimental
+}  // namespace Kokkos
+
+namespace Kokkos {
+namespace Experimental {
+
+class ViewHolderBase {
+ public:
+  virtual size_t span() const                      = 0;
+  virtual bool span_is_contiguous() const          = 0;
+  virtual void *data() const                       = 0;
+  virtual void *rec_ptr() const                    = 0;
+  virtual std::string label() const noexcept       = 0;
+  virtual bool is_data_type_const() const noexcept = 0;
+
+  virtual ViewHolderBase *clone() const      = 0;
+  virtual size_t data_type_size() const      = 0;
+  virtual bool is_hostspace() const noexcept = 0;
+
+  // the following are implemented in specialization classes.
+  // View Holder is only a pass through implementation
```

Comment:
Should this comment be moved up? It looks like it's true for span and the rest above